### DISCUSSION TITLE
next and previous playback control support for webradios

### DISCUSSION THREAD

**Discussion by pirateradiohack** [jcorporation/myMPD#935] [Q&A]
I'm not sure to understand what is the best way to skip to play the next favorite webradio.

This might not be a big issue from within the web interface as we can switch to another webradio by selecting it in the interface. But regarding physical buttons on a device I find it very convenient to be able to skip to the next webradio.

I understand those playback controls have an effect in the current queue.

I have a list of favorite webradios in the browse -> view -> webradios.

- When I click the play button within the favorite icon then it replaces the current queue and plays this specific webradio. And so using the next or previous playback control has no effect.

- I could add all the favorite webradios to the current queue but I don't know if there is a simple way to achieve this (all I could find is the individual add to queue in the context menu, so one by one).
And besides that, I would have to update the current queue with any new favorite webradios. Quite cumbersome. 

- I could not find an easy way to use the myMPD API to achieve this either. I guess I could use the endpoint that retrieves the list of favorite webradios and hack something based on that but I'd rather avoid it if there is a better solution.

Any guidance on the subject is appreciated.

**Comment by jcorporation** [comment]
I think this should be doable by a lua script.

A script that:

1. retrieves the current playing webradio
2. gets all webradio favorites
3. finds the current playing webradio
4. selects the next webradio
5. replaces the queue

**Reply by pirateradiohack** [reply]
Yes, I guess that's the way to go. 

I feel more comfortable with Python but I can try my hands on Lua.

**Reply by jcorporation** [reply]
Lua can be directly executed inside the myMPD process: https://jcorporation.github.io/myMPD/scripting/. Be ware that the scripting api will change in the v10.2.0 release.

**Reply by pirateradiohack** [reply]
Shall I expect my script to break by v10.2.0?

**Reply by pirateradiohack** [reply]
I came to realize that there is a fully functional IDE within the web interface. Another incredible feature!

**Reply by jcorporation** [reply]
Yes, there are some breaking changes, but this changes can be easily adapted.

Bigest change is:

Old:
```
rc, raw_result = mympd_api_raw("MYMPD_API_PLAYLIST_LIST", json.encode({offset = 0, limit = 2000, searchstr = "", type = 0}))
result = json.decode(raw_result)
```

New:
```
rc, result = mympd.api("MYMPD_API_PLAYLIST_LIST", {
    offset = 0,
    limit = 2000,
    searchstr = "",
    type = 0
})
```

You can have a look at the example scripts: https://github.com/jcorporation/myMPD/commits/v10.2.0/docs/scripting/scripts

**Reply by pirateradiohack** [reply]
any recommendation as to how to get the properties of a favorite? a place where I could peak at the model?

**Reply by pirateradiohack** [reply]
specifically I'm looking for some sort of ID. Does that even exist? or should refer to the URI?
Are those the properties of the object?  https://github.com/jcorporation/myMPD/blob/6392a31e0ff570b42a11f148d00523e024ec8b3b/src/mympd_api/webradios.c#L218

**Reply by jcorporation** [reply]
The filename is the uniq identifier and is a derivate of the stream uri, but uri should work also.

From WebradioDB documentation:

> The filename of the playlist and the coverimage are derived from the streamuri by replacing <>/.:?&$!#\|;= characters with _. This is the same behaviour as in myMPD and makes this playlists compatible with the myMPD webradio feature.

The easiest way to explore the json rpc api of myMPD is to build and run myMPD in debug mode, and append `/test/manual.html` to the uri, e.g. `http://localhost:8080/test/manual.html`. This is an undocumented feature ;)

![image](https://user-images.githubusercontent.com/765844/208313295-332d9e9f-2378-4571-af4a-8c348ca00fcd.png)

**Reply by pirateradiohack** [reply]
```Lua
rc, raw_result = mympd_api_raw("MYMPD_API_WEBRADIO_FAVORITE_LIST", json.encode({offset = 0, limit = 50, searchstr = "tabula"}))
if rc == 0 then
    result = json.decode(raw_result)
end

print(result)
```

returns

`{"jsonrpc":"2.0","id":0,"result":{"method":"MYMPD_API_SCRIPT_EXECUTE","facility":"script","severity":"info","message":"ok","data":{}}}`

Does that mean the endpoint returns an empty table for the webradio favorite list? Or did I do something wrong?

**Reply by jcorporation** [reply]
This means, that the script was executed successfully, look at the `method`.

The `print(result)` output is printed to `stdout` of the myMPD process. If you want to return any text as notification to the myMPD webui, you should use `return(result)`.

**Reply by pirateradiohack** [reply]
I'm executing it in the shell with `mympd-script` and was hoping print would return something. 
I'm trying to get the webradio favorite list, will try again :)

**Reply by jcorporation** [reply]
The script is executed asynchronous in a separate thread. That is the reason why myMPD-script can not return the script output.

**Reply by pirateradiohack** [reply]
I'm improving my Lua skills by the minute...

So, I thought this was the way to check how many favorites web radios there are. But it returns 0, and I have more than 0 favorite web radios. 

```Lua
rc, raw_result = mympd_api_raw("MYMPD_API_WEBRADIO_FAVORITE_LIST", json.encode({offset = 0, limit = 50, searchstr = "tabula"}))
if rc == 0 then
    radios_list = json.decode(raw_result)
end

return(radios_list.result.totalEntities)
```

I guess I missed something. Will try again.

**Reply by jcorporation** [reply]
You search for a webradio with „tabula“ in the name, use a blank searchstr to get all favorites.

**Reply by pirateradiohack** [reply]
wow such an improvement :) thanks. I might be able to get somewhere.

**Reply by pirateradiohack** [reply]
The API documentation is nice, but without the models I find it quite difficult to work with it. 
(By the way, the link given in the IDE has probably moved, it's a 404 for me: https://jcorporation.github.io/mympd/apidoc.html)
I guess this is the API doc https://jcorporation.github.io/myMPD/references/api/methods. 
Please let me know if there is a place where I can check on the models.

At the moment I would like to peak at the data returned by `MYMPD_API_PLAYER_CURRENT_SONG` in order to look for a common unique key with `MYMPD_API_WEBRADIO_FAVORITE_LIST`.

Thanks!

**Comment by pirateradiohack** [comment]
For the moment I find it easier to use Lua to load all the favorite radios into the queue:

```Lua
-- retrieve list of favorite radios
rc, raw_result = mympd_api_raw("MYMPD_API_WEBRADIO_FAVORITE_LIST", json.encode({offset = 0, limit = 50, searchstr = ""}))
if rc == 0 then
    radios_list = json.decode(raw_result)
else
    return("Failed to retrieve the list of favorite radios.")
end

-- clear the queue
rc, raw_result = mympd_api_raw("MYMPD_API_QUEUE_CLEAR", json.encode({}))
if rc == 0 then
    result = json.decode(raw_result)
else
    return("Failed to clear the queue.")
end

-- add each of favorite URI to the queue
for i, radio in ipairs(radios_list.result.data) do
    rc, raw_result = mympd_api_raw("MYMPD_API_QUEUE_APPEND_URI", json.encode({uri = radio.StreamUri, play = false}))
    if rc == 0 then
        result = json.decode(raw_result)
    else
        return("Failed to append uri to queue")
    end
end

-- start playing the radios
rc, raw_result = mympd_api_raw("MYMPD_API_PLAYER_PLAY", json.encode({}))
if rc == 0 then
    result = json.decode(raw_result)
else
    return("Failed to play the queue.")
end
```

That has the advantage of keeping the MPD standard use of the next and previous playback control buttons.

Although I wish I could set the repeat option of MPD to true.  
But the example API call:

```Lua
-- make sure the queue is in repeat mode so reaching the end will jump back to the beginning
rc, raw_result = mympd_api_raw("MYMPD_API_PLAYER_OPTIONS_SET", json.encode({consume = "1", random = false, single = "1", repeat = false, replaygain = "off", crossfade = 0, mixrampDb = 0, mixrampDelay = 0, jukeboxMode = "off", jukeboxPlaylist = "Database", jukeboxQueueLength = 1, jukeboxLastPlayed = 24, jukeboxUniqueTag = "Album", autoPlay = false}))
if rc == 0 then
    result = json.decode(raw_result)
end
```

seems to return an error:
`Error executing script Play Radios: Syntax error during precompilation: /var/lib/mympd/scripts/Play Radios.lua:37: unexpected symbol near 'repeat'`

Any idea what could be wrong here? Maybe `repeat` is a Lua reserved keyword?

**Reply by jcorporation** [reply]
Correct, `repeat` is a reserved keyword and I know no method to escape it.

Use:

```Lua
options = {}
options["repeat"] = true

rc, raw_result = mympd_api_raw("MYMPD_API_PLAYER_OPTIONS_SET", json.encode(options))
if rc == 0 then
    result = json.decode(raw_result)
end
```

This syntax will be the default in v10.2.0.
